zeno/agents/docfinder/utils/nodes.py

```diff
@@ -45,15 +42,21 @@ class grade(BaseModel):
         input_variables=["context", "question"],
     )
 
+    model_id = config["configurable"].get("model_id")
+    model = ModelFactory().get(model_id)
```

Comment:
Not sure how much overhead this is to instanciate every time. Maybe we need to pull these out of the tools and instanciate on module load? Something that does not need change right away but to keep in mind.

Reply:
Good call. I think there are 3 ways to address this: 
- Instantiate the user-specified model and store is as an environment variable, before running the `graph.stream` command. Each tool can then pull the model from that environment variable. I'm not a huge fan of using environment variables in this way. 
- Instantiate the user-specified model and store it in the `config` object passed to the `graph.stream()` method. (I'm not sure what the restrictions are on the values in the config object - they may need to be hashable for example)
- Possibly an easier options is to add a `functools.lru_cache` to the `ModelFactory.get()` method to cache the instantiated model. Since the method output is cached based on the hashed parameters, the cache would maintain one version of the model for JSON mode and one for regular mode for each model

Reply:
As discussed earlier, we'll leave the model instantiation as is for now, as we may want to explore the option of the tools/nodes using different models